main/WebServer.cpp

```diff
@@ -16629,1342 +16656,1517 @@ namespace http {
+                        int intGust = atoi(sd[4].c_str());
+                        if (m_sql.m_windunit != WINDUNIT_Beaufort)
+                        {
+                            sprintf(szTmp, "%.1f", float(intSpeed) * m_sql.m_windscale);
+                            root["result"][ii]["sp"] = szTmp;
+                            sprintf(szTmp, "%.1f", float(intGust) * m_sql.m_windscale);
+                            root["result"][ii]["gu"] = szTmp;
+                        }
+                        else
+                        {
+                            float windspeedms = float(intSpeed) * 0.1F;
+                            float windgustms = float(intGust) * 0.1F;
+                            sprintf(szTmp, "%d", MStoBeaufort(windspeedms));
+                            root["result"][ii]["sp"] = szTmp;
+                            sprintf(szTmp, "%d", MStoBeaufort(windgustms));
+                            root["result"][ii]["gu"] = szTmp;
+                        }
+                        ii++;
+                    }
+                }
+            }//custom range
+        }
+
+        /**
+         * Retrieve user session from store, without remote host.
+         */
+        WebEmStoredSession CWebServer::GetSession(const std::string &sessionId)
+        {
+            //_log.Log(LOG_STATUS, "SessionStore : get...");
+            WebEmStoredSession session;
+
+            if (sessionId.empty()) {
+                _log.Log(LOG_ERROR, "SessionStore : cannot get session without id.");
+            }
+            else {
+                std::vector<std::vector<std::string> > result;
+                result = m_sql.safe_query("SELECT SessionID, Username, AuthToken, ExpirationDate FROM UserSessions WHERE SessionID = '%q'",
+                        sessionId.c_str());
+                if (!result.empty()) {
+                    session.id = result[0][0];
+                    session.username = base64_decode(result[0][1]);
+                    session.auth_token = result[0][2];
+
+                    std::string sExpirationDate = result[0][3];
+                    //time_t now = mytime(NULL);
+                    struct tm tExpirationDate;
+                    ParseSQLdatetime(session.expires, tExpirationDate, sExpirationDate);
+                    // RemoteHost is not used to restore the session
+                    // LastUpdate is not used to restore the session
+                }
+            }
+
+            return session;
+        }
+
+        void CWebServer::GroupBy(
+                Json::Value &root,
+                std::string dbasetable,
+                uint64_t idx,
+                std::string sgroupby,
+                std::function<std::string (std::string)> counterExpr,
+                std::function<std::string (std::string)> valueExpr,
+                std::function<std::string (double)> sumToResult
+        ) {
+            std::string query = boost::str(boost::format(std::string("")
```

Comment:
This could be a normal std::string, but the safe_query function also accepts parameters and is much clearer

Reply:
I'm cross-compiling on WSL2, in a Debian distro. It works very well, but I guess some things are not easy to emulate. I'm cross-compiling for a RPi3, so I do need to have the GPIO code in the executable.

Reply:
Thanks for pointing it out, the function std_format! I hadn't seen it before. I made a similar function and called it StdFormat ;-) Yours is much better so I'm using it, of course. For the boost::format, I made an alternative that replaces positional placeholders; I find it much better to have the arguments just passed once and be able to refer to them by index. It's something the Java-equivalent of printf can do, but unfortunately in C++ this was never implemented.

Reply:
I am glad it's not in c++ ;) ... It's clearer to use the safe_query function
Is it possible to use 'std_format' instead of StdFormat, No need for new functions, will also make the application bigger and could confuse feature developers... which function to use...